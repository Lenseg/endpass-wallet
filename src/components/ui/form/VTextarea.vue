<template>
  <div class="field">
    <label class="label"
           v-if="label"
           :for="id">{{ label }}</label>
    <div class="control">
      <textarea
        v-model="innerValue"
        class="textarea"
        v-bind="getTextareaProps()"
      />
    </div>
  </div>
</template>

<script>
export default {
  name: 'v-textarea',
  props: {
    value: {
      type: [String, Number],
      default: null,
    },
    id: {
      type: String,
      default: null,
    },
    label: {
      type: String,
      default: null,
    },
    placeholder: {
      type: String,
      default: null,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
  },
  data() {
    return {
      innerValue: this.value,
      ignoreProps: ['label'],
    };
  },
  methods: {
    getTextareaProps() {
      return Object.keys(this.$props).reduce((accumulator, prop) => {
        if (!this.ignoreProps.includes(prop)) {
          accumulator[prop] = this.$props[prop];
        }

        return accumulator;
      }, {});
    },
  },
  watch: {
    value(value) {
      this.innerValue = value;
    },
    innerValue(value) {
      this.$emit('input', value);
    },
  },
};
</script>
