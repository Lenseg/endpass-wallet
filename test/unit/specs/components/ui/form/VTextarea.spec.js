import { shallow } from '@vue/test-utils';

import VTextarea from '@/components/ui/form/VTextarea';

describe('VTextarea', () => {
  describe('render', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(VTextarea);
    });

    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('v-textarea');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render component with label', () => {
      wrapper.setProps({ label: 'Label' });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('props', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(VTextarea);
    });

    it('should render textarea props', () => {
      const props = {
        id: 'id',
        value: 'value',
        placeholder: 'placeholder',
        disabled: 'disabled',
        label: 'label',
      };
      const expectedProps = {
        id: 'id',
        placeholder: 'placeholder',
        disabled: 'disabled',
      };

      wrapper.setProps(props);

      expect(wrapper.find('textarea').attributes()).toEqual(
        expect.objectContaining(expectedProps),
      );
    });

    describe('value', () => {
      it('should change textarea value', () => {
        let value = 'value';

        wrapper.setProps({ value });
        expect(wrapper.find('textarea').element.value).toBe(value);

        value = '';
        wrapper.setProps({ value });
        expect(wrapper.find('textarea').element.value).toBe(value);
      });
    });
  });
});
