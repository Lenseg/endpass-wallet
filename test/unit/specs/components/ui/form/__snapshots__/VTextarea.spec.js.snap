// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTextarea render should render component 1`] = `
<div
  class="field"
>
  <!---->
   
  <div
    class="control"
  >
    <textarea
      class="textarea"
    />
  </div>
</div>
`;

exports[`VTextarea render should render component with label 1`] = `
<div
  class="field"
>
  <label
    class="label"
  >
    Label
  </label>
   
  <div
    class="control"
  >
    <textarea
      class="textarea"
    />
  </div>
</div>
`;
